### ISSUE TITLE
Worldgen Replacement Issues

### ISSUE THREAD

**Issue by ghost** [Delocuro/Minestrapp-1.12#15]
Minecraft Version: 1.12.2
Forge Build: 2611
Minestrapp Version: 5.2.1A
[Ocean Floors](http://minecraft.curseforge.com/projects/ocean-floor-clay-sand-and-dirt?projectID=231837) Version: 1.0.1
Picture: https://imgur.com/a/c7ln0

When generating ocean floors, or the lake beds pictured, generation seems to halt at chunk edges. I've noticed the same thing happens with the Minestrapp ore replacements for the individual stone types; any ore for that vein that generates in a different chunk, even if it's the same biome, gets the vanilla stone type, rather than the biome stone type.

A bit of good news, however, is that CofH World _is_ compatible with the ore replacement that Minestrapp does (vanilla ore types spawned by CofH World get converted, but still suffer stone type bugs at chunk borders)!

**Comment by Delocuro** [comment]
Yeah, that's a known bug that I'm still not totally sure how to fix. I know *why* it happens - when features such as the sand circles seen in that screenshot generate in new chunks and extend into old ones, they check to see if there are dirt blocks they can generate in - the new chunk has them since Minestrapp's block replacements usually happen last, but since the old chunks have already been replaced, they no longer have Vanilla dirt, and thus the circles get cut off.

What I'm not sure of is how to fix it. Werty added a 1/2 chunk buffer to all the Minestrapp generation methods, which solved some of the problems, but it still seems to happen seemingly at random during worldgen, usually when generating lots of chunks quickly (i.e., when sprint-flying into unexplored areas in Creative). One thing I'd considered is that maybe we need to reverse the chunk buffer when generating blocks in the opposite direction of spawn, but I haven't been able to find consistent evidence to support that.

If you're able, try to find a Jungle on the opposite side of the map (if you're at positive x, positive z, try finding one at negative x, negative z, for instance) and see if the issue seems to be happening there too. That's a much more visually obvious example of the bug than most instances I've encountered with just Minestrapp installed, so maybe you'll be able to find some correlation that I haven't been able to see yet...

**Comment by ghost** [comment]
Same album link, but updated with new pictures; https://imgur.com/a/c7ln0

I've found that it does indeed occur on the other side of the world coordinates with biomes that replace vanilla dirt with Minestrapp dirt. In the second (new) image, it's a little hard to see under the ice, but it's definitely happening along chunk borders there as well.

**Comment by Delocuro** [comment]
Hmmm... guess it isn't an issue with the buffer values, then. I've been trying to get this stupid bug fixed since before the mod was released, but I've had no luck so far, so I guess I'll have to look into some other 1.12 worldgen mods and see if I can find anything they're doing that I missed...

**Comment by ghost** [comment]
I read somewhere about how some structures that span multiple chunks don't actually spawn in their blocks until all the chunks they'll occupy have been generated. Perhaps you could do the same thing with the features that are getting cut off, or not changed to the right stone type, and not generate them in a chunk until all the surrounding chunks have been generated? I'm not sure how feasible this idea is though.

**Comment by Delocuro** [comment]
Don't think this'll be fixed for 5.3.0 unfortunately. I've looked into solving the issue for a number of other features, but I'm really not sure how to go about it. I'll leave it open for a future update -- it's an issue that definitely needs fixing, I just don't know how to solve it yet.

**Comment by Delocuro** [comment]
Renamed since it's a broader issue than just mod support -- hopefully we can figure this out for 5.4.0
